File: mini-mart/src/main/java/com/example/mini_mart/entity/Product.java
package com.example.mini_mart.entity;

import jakarta.persistence.*;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Size;
import lombok.Getter;

import lombok.Setter;
import org.hibernate.annotations.ColumnDefault;

import java.math.BigDecimal;
import java.time.Instant;
import java.util.ArrayList;
import java.util.List;

@Getter
@Setter
@Entity
@Table(name = "product")
public class Product {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id", nullable = false)
    private Integer id;

    @NotNull
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "category_id", nullable = false)
    private Category category;

    @NotNull
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "supplier_id", nullable = false)
    private Supplier supplier;

    @Size(max = 50)
    @NotNull
    @Column(name = "sku", nullable = false, length = 50)
    private String sku;

    @Size(max = 500)
    @NotNull
    @Column(name = "name", nullable = false, length = 500)
    private String name;

    @Size(max = 250)
    @Column(name = "brand", length = 250)
    private String brand;

    @Lob
    @Column(name = "description")
    private String description;

    @Size(max = 500)
    @Column(name = "image_url", length = 500)
    private String imageUrl;

    @Size(max = 50)
    @Column(name = "unit", length = 50)
    private String unit;

    @NotNull
    @Column(name = "selling_price", nullable = false, precision = 18, scale = 2)
    private BigDecimal sellingPrice;

    @NotNull
    @ColumnDefault("1")
    @Column(name = "is_active", nullable = false)
    private Boolean isActive = false;

    @NotNull
    @ColumnDefault("CURRENT_TIMESTAMP")
    @Column(name = "created_at", nullable = false)
    private Instant createdAt;

    @PrePersist
    void prePersist() {
        this.createdAt = Instant.now();
    }

    @OneToMany(mappedBy = "product")
    private List<ExportReceiptItem> exportReceiptItems = new ArrayList<>();

    @OneToMany(mappedBy = "product")
    private List<ImportReceiptItem> importReceiptItems = new ArrayList<>();

}

File: mine-mart-fe/src/types/product.ts
/**
 * Product types - maps to backend DTOs
 */

export interface ProductResponse {
  id: number;
  categoryId: number;
  categoryName?: string;
  supplierId: number;
  supplierName?: string;
  sku: string;
  name: string;
  brand?: string;
  description?: string;
  imageUrl?: string;
  unit?: string;
  sellingPrice: number;
  isActive: boolean;
  createdAt: string;
}

export interface ProductCreateRequest {
  categoryId: number;
  supplierId: number;
  sku: string;
  name: string;
  brand?: string;
  description?: string;
  imageUrl?: string;
  unit?: string;
  sellingPrice: number;
}

export interface ProductUpdateRequest {
  categoryId?: number;
  supplierId?: number;
  sku?: string;
  name?: string;
  brand?: string;
  description?: string;
  imageUrl?: string;
  unit?: string;
  sellingPrice?: number;
  isActive?: boolean;
}

